### PR TITLE
Use state to update a Command's input.

### PR DESCRIPTION
**Pull request by ChickenProp**
This allows you to write commands whose input changes when a previous
Command shrinks.

For example, suppose the state contains `someList :: [Bool]` and
you have a command whose input expects "index into `someList` pointing
at `True`". You can generate that index directly, but if an earlier
command shrinks, `someList` might change and your index now points at
`False`.

(This is contrived, but hopefully points at the sorts of more
complicated situations where it might be useful.)

With this you can instead generate a number between 0 and (the number of
`True` elements). Then use `mkInput` to turn that number into an index
into `someList`. This will still be valid as long as the number of
`True` elements doesn't shrink below the generated value.

You could also pass this number directly into `exec`. But then in `exec`
you'd need to get `someList` directly from the concrete model, which
might be complicated and/or slow.

I implemented this by adding a new `Command` constructor, `CommandA`
where A is for Advanced. I don't love this. I could have simply changed
the existing constructor, but that means every existing Command needs to
be updated. (It's a simple change, `commandMkInput = const Just` means
they work as before, but still a massive pain.) The downside of this
approach is implementation complexity, plus any user functions taking a
`Command` as input may need to be updated.

Other approaches we could take here:

1. We could pass the concrete state into `exec` along with the concrete
input. Then you wouldn't need to get `someList` from the model. But you
might still need to do complicated calculations in `exec` which could
make the failure output hard to follow.

     If we did this we'd have the same tradeoff between changing the
     existing constructor and adding a new one.

2. We could add a callback

    ```haskell
    MapMaybeInput
      (state Symbolic -> input Symbolic -> Maybe (input Symbolic)
    ```

    Each of these would be applied in turn to update the input.
    (`Require` would be equivalent to a `MapMaybeInput` that ignores
    state, and either returns the input unchanged or `Nothing`.)

    This would be compatible with existing commands, but functions
    accepting a `Command` or a `Callback` as input might still need
    changing.

### CODE REVIEW

**Comment by ChickenProp** [comment]
If we think this works for us, I'll make a PR upstream. It depends on #1 (https://github.com/hedgehogqa/haskell-hedgehog/pull/453). I've emailed the maintainer, so I'm hoping he'll be able to discuss, because I'm not 100% confident this gets all the edge cases.

File: hedgehog/src/Hedgehog/Internal/State.hs [hunk 1]
```diff
@@ -512,26 +555,28 @@ contextNewVar = do
   put $ Context state (insertSymbolic var vars)
   pure var
 
--- | Drops invalid actions from the sequence.
+-- | Pass the state through the actions, updating inputs and dropping invalid
+--   ones.
 --
-dropInvalid :: [Action m state] -> State (Context state) [Action m state]
-dropInvalid =
+rethreadState :: [Action m state] -> State (Context state) [Action m state]
```
Comment (mgajda) [inline hedgehog/src/Hedgehog/Internal/State.hs:558]:
```
-- | Here we re-run the thread of all actions within the action list to assure all actions remain valid.
--   This avoids repeating the rethreading logic in each monadic action DSL.
```

**Review by mgajda — APPROVED** [review]

**Comment by mgajda** [comment]
This is very nice standard facility, but I think its description should be copied from the patch to the module, since it is non-obvious how it works.

File: hedgehog/src/Hedgehog/Internal/State.hs (hunk 1 above)
Comment (ChickenProp) [reply hedgehog/src/Hedgehog/Internal/State.hs:558]: I confess I think that comment is less clear than the current one. I don't know what a "monadic action DSL" refers to - is it simply an `Action` data type? But the rethreading logic is something that has to happen with a list of actions, not individual ones.

**Comment by ChickenProp** [comment]
I feel like a lot of what I've written here wouldn't be very good code documentation - there's a bunch of historical notes and "here are some ways this code could work but doesn't". It feels more like it belongs in a tutorial or something. (Maybe a tutorial kept in the codebase, but I don't think there's a good place for it currently.)

File: hedgehog/src/Hedgehog/Internal/State.hs (hunk 1 above)
Comment (mgajda) [reply hedgehog/src/Hedgehog/Internal/State.hs:558]: Yes, it refers to `Action` datatype.
But you're right.
I will propose comments where they are missing, and we can talk about these.

File: hedgehog/src/Hedgehog/Internal/State.hs [hunk 2]
```diff
@@ -661,7 +658,7 @@ newtype Sequential m state =
     }
 
 renderAction :: Action m state -> [String]
```
Comment (mgajda) [inline hedgehog/src/Hedgehog/Internal/State.hs:660]:
```
-- | Shows input and output variables of the action.
```

File: hedgehog/src/Hedgehog/Internal/State.hs [hunk 3]
```diff
@@ -677,7 +674,7 @@ renderAction (Action _ input _ (Symbolic (Name output)) _ _ _ _) =
         fmap (prefix ++) xs
 
 renderActionResult :: Environment -> Action m state -> [String]
-renderActionResult env (Action _ _ _ output@(Symbolic (Name name)) _ _ _ _) =
+renderActionResult env (Action _ _ output@(Symbolic (Name name)) _ _ _ _) =
```
Comment (mgajda) [inline hedgehog/src/Hedgehog/Internal/State.hs:677]:
```
-- | Render the action, telling if its output is not matching the environment.
```

File: hedgehog/src/Hedgehog/Internal/State.hs [hunk 4]
```diff
@@ -776,7 +773,7 @@ data ActionCheck state =
     }
```
Comment (mgajda) [inline hedgehog/src/Hedgehog/Internal/State.hs:773]:
```
-- | Re-validations needed for a given `Action` after a shrink.
```

File: hedgehog/src/Hedgehog/Internal/State.hs [hunk 5]
```diff
@@ -449,17 +449,14 @@ commandGenOK (CommandA inputGen _ _ _) state =
 --   evaluated.
 --
 data Action m (state :: (Type -> Type) -> Type) =
-  forall input0 input output.
+  forall input output.
   (TraversableB input, Show (input Symbolic), Show output) =>
   Action {
```
Comment (mgajda) [inline hedgehog/src/Hedgehog/Internal/State.hs:454]: Am I correct that we miss the name of the action here?
This would be purely for debugging purposes, but may make it easier to see what happens.

File: hedgehog/src/Hedgehog/Internal/State.hs (hunk 4 above)
Comment (mgajda) [reply hedgehog/src/Hedgehog/Internal/State.hs:773]: I also wonder if `ctx1 { contextState = contextState ctx0 }` is the best way of describing what happens.
I understand that we generate commands with gen `commands`  from state in `ctx0`, in both branches.
Why would we use `ctx1` to start the second branch? Is it to avoid using the same variable names?|

If so, it is a hidden interaction that merits a justification in a comment.

Comment (ChickenProp) [reply hedgehog/src/Hedgehog/Internal/State.hs:773]: That sounds right, yeah. I haven't looked at that code in any depth. I don't want to add a bunch of comments to code that I haven't taken the time to fully understand, especially not in a PR that doesn't touch that code.
